### ISSUE TITLE
Fix apache2 service failing to start in DO

### ISSUE THREAD

**Issue by conorsch** [freedomofpress/securedrop#1078]
Provisioning runs recently started failing in DigitalOcean droplets. Specifically apache2 fails to start when triggered in the post-install scripts during installation of the `securedrop-app-code` package. The that fails is `ensure securedrop-app-code package is installed from repo`. The interesting part of the error output is: 

```
Cleaning up...
+ chown -R www-data:www-data /var/www/securedrop
+ chown www-data:www-data /var/www/document.wsgi
+ chown www-data:www-data /var/www/source.wsgi
+ a2dissite 000-default
Site 000-default already disabled
+ a2dissite default-ssl
Site default-ssl already disabled
+ service apache2 reload
 * Reloading web server
 * 
+ [ -e /etc/apparmor.d/disable/usr.sbin.apache2 ]
+ aa-enforce /etc/apparmor.d/usr.sbin.tor
Setting /etc/apparmor.d/usr.sbin.tor to enforce mode.
+ aa-enforce /etc/apparmor.d/usr.sbin.apache2
Setting /etc/apparmor.d/usr.sbin.apache2 to enforce mode.
+ service apache2 restart
 * Restarting web server 
                                        [fail]
dpkg: error processing package securedrop-app-code (--configure):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
 securedrop-app-code
E: Sub-process /usr/bin/dpkg returned an error code (1)
```
Full error output visible [in Snap](https://snap-ci.com/conorsch/securedrop/branch/snapci-serverspec-prod-tests/logs/defaultPipeline/12/Staging_Clean). Logging into the server after the failure and manually kicking the service also fails:

```
root@app-staging:~# time service apache2 restart
 * Restarting web server apache2               [fail] 

real    1m0.272s
user    0m0.061s
sys     0m0.158s
```
Cannot reproduce in VirtualBox, so focus on DO.

**Comment by conorsch** [comment]
Forcing installation of `securedrop-app-code=0.3.3` (rather than `0.3.4`) prevents the failure.

**Comment by dolanjs** [comment]
Don't know why these lines/permissions aren't needed on physical or virtualbox but are needed in DO.

```
  signal (send) set=("term") peer=unconfined,
  ptrace (trace) peer=unconfined,
```

**Comment by garrettr** [comment]
Those are the lines that were removed from the AppArmor profile in #1058 as part of the resolution for one of the issues from the last iSec audit.

**Comment by conorsch** [comment]
The iSec audit only recommended removing `capability dac_override`. Removing the extra lines was overzealous, but admittedly didn't cause any problems in many iterations of testing in virtualbox, and even in hardware.

Turns out at least the `signal` declaration is required for running apache2 on DigitalOcean hosts. Right now my best guess is that the requirement for signal passing between unprofiled (and therefore unconfined) peers is due to container-based virtualization for the provider. The `ptrace`  declaration didn't show up in aa-complain mode during a quick check I just ran, but I'm going to run some tests on DO hosts and see if I can find it again via aa-logprof.

For reference, the lines in question were added to the AppArmor profile for apache2 in:

* b6f97d3f: `signal`, `dac_override`
* b6fcd8a5: `ptrace`, `sys_ptrace`

After some more testing I'm going to recommend adding add least the `signal` declaration back in, possibly both, since fixing provisioning on DO hosts is currently blocking work on the automated testing suite (#1070).

**Comment by garrettr** [comment]
> Right now my best guess is that the requirement for signal passing between unprofiled (and therefore unconfined) peers is due to container-based virtualization for the provider.

Can we do better than a guess on why this is needed for DO, but not for Virtualbox or hardware?

**Comment by garrettr** [comment]
Originally, we thought this error was specific to DigitalOcean, since we could only reproduce it there (and not on Virtualbox or hardware). Today @conorsch realized that the key factor is the use of a grsecurity-enabled kernel, which is used by default in both Virtualbox and hardware installations but is _not_ used on DigitalOcean because DO does not support custom kernels (due to their virtualization setup).

You can reproduce the error locally in Virtualbox **if you skip the `grsec` tag** for the staging or prod roles and `vagrant up` either environment.

The failure occurs during the `ensure securedrop-app-code package is installed from repo` task. Near the end of its run, the `securedrop-app-code` postinst script sets the apparmor scripts for tor and apache to enforce mode and then restarts both processes to ensure they are running with enforcement. On a non-grsecurity host, apache fails to restart which causes the install process to fail:

```sh
<snip>
Setting /etc/apparmor.d/usr.sbin.tor to enforce mode.
Setting /etc/apparmor.d/usr.sbin.apache2 to enforce mode.
 * Restarting web server apache2
   ...fail!
<snip>
```

We have already identified these AppArmor lines as necessary for it to work without grsecurity:

```
signal (send) set=("term") peer=unconfined,
ptrace (trace) peer=unconfined,
```

## Why does it work without grsecurity?

Signal and ptrace mediation is a [relatively new feature](https://penguindroppings.wordpress.com/2014/06/06/application-isolation-with-apparmor-part-iv/) in AppArmor and *requires kernel support*, which was only introduced in Ubuntu 14.04 LTS. From the [Launchpad bug](https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1298611):

> When used with a compatible kernel, signals and ptrace rules are supported. When used without a compatible kernel (eg, on Ubuntu Touch for a few weeks or with upstream kernels), signal and ptrace rules are skipped (ie, you can use this userspace with other kernels without issue).

The grsecurity kernel is just upstream+grsecurity, and only uses the package scripts from the Ubuntu kernel overlay, so it does not include the necessary kernel support for signal and ptrace mediation. As a result, when these AppArmor profiles are used with a grsecurity kernel, the corresponding rules are skipped and AppArmor does not interfere with Apache restarts.

## Why are these rules needed for Apache to restart successfully?

### signal

```
signal (send) set=("term") peer=unconfined
```

This allows the apache2 process to send SIGTERM to any unconfined peer (another process without AppArmor restrictions). If you think about it, this makes sense - when Apache restarts, it needs to send SIGTERM to all of the *child worker processes* it spawned. 

This was noticed upstream [here](https://bugs.launchpad.net/ubuntu/+source/apparmor/+bug/1322764) and fixed [here](https://bazaar.launchpad.net/~apparmor-dev/apparmor/master/revision/2533), where you can see the addition of a `signal (send)` to resolve the issue.

If you examine `/var/log/syslog` after attempting to restart apache *without* the signal rule, you will see a corresponding DENY:

```
Jul 29 02:18:38 app-staging kernel: [  185.830734] type=1400 audit(1438136318.126:29): apparmor="DENIED" operation="signal" profile="/usr/sbin/apache2" pid=8473 comm="apache2" requested_mask="send" denied_mask="send" signal=term peer="unconfined"
```

I tested this by manually adding the `signal` rule to `/etc/apparmor.d/usr.sbin.apache2`, running `aa-disable usr.sbin.apache2 && aa-enforce usr.sbin.apache2` so it would pick up the change, and then `service apache2 restart`, which succeeded without errors.

Note that we may be able to use a more restrictive signal rule than one that simply allows apache to send SIGTERM to all unconfined peers -- we should file a follow-up bug.

### ptrace

```
ptrace (trace) peer=unconfined
```

I don't think apache needs to ptrace any processes. In my testing, I only needed to re-add the `signal` to allow Apache to restart again. Additionally, it is not present in the [latest version](https://bazaar.launchpad.net/~apparmor-dev/apparmor/master/view/head:/profiles/apparmor.d/usr.sbin.apache2) of the Apache AppArmor rules that are included in Ubuntu.

We should leave this rule out unless we find a compelling reason to include it.

**Comment by conorsch** [comment]
@garrettr Great write-up. I'm responding here rather than in #1089 so the discussion stays readable.

The `peer=unconfined` is only necessary when apache2 starts without the apparmor profile enabled. We should be able to tweak the `postinst` script in the `securedrop-app-code` package to handle enforcement correctly across hosts and platforms. 

Regarding the `ptrace` permission, I noticed `DENIED` entries for ptrace when running the app tests in staging. There's definitely a case to be made that we shouldn't ship additional permissions to prod in order to provide functionality in staging, and we should have that conversation. First, though, I want to see a more concrete list of actions that generate `ptrace` requests. A few ideas to check:

* an exception is generated in the app code
* successful communication to redis, gpg, or scrypt
* falsey communication to redis, gpg, or scrypt

What's most disconcerting about finding this error is:

> when these AppArmor profiles are used with a grsecurity kernel, the corresponding rules are skipped and AppArmor does not interfere with Apache restarts.

What else is silently failing?

**Comment by conorsch** [comment]
As explained above, the ostensibly DO-specific failures are actually related to the absence of the grsecurity kernel patches, and consequently how AppAmor behaves with or without grsecurity present. One possible resolution is re-adding the `signal` declaration, as proposed in #1089. In the spirit of the principle of least privilege, however, we should avoid granting any unnecessary privileges if at all possible (even if we were overzealous in removing them in the first place).

The problem with `signal (send) set=("term") peer=unconfined` is that it permits the apache2 process to send SIGTERM to any other process on the system, as long as the receiving process does not have an AppArmor profile—which effectively includes every process except for tor on SecureDrop servers. If apache2 itself starts unconfined, but the profile is later enforced while the process is running, then apache2 will fail to restart, since it would need to kill its unconfined ancestor before starting a new confined process, and the enforced profile doesn't permit that without `signal (send) set=("term") peer=unconfined`. (Interesting aside: `service apache2 restart` sends SIGTERM, but `apache2ctl restart` sends SIGHUP.) The postinst script in the securedrop-app-code package tries to enforce policy on a running apache2 process:

```
    service apache2 reload

    # If the profile was disabled enabled it.
    if [ -e "/etc/apparmor.d/disable/usr.sbin.apache2" ]; then
        rm /etc/apparmor.d/disable/usr.sbin.apache2
    fi

    aa-enforce /etc/apparmor.d/usr.sbin.tor
    aa-enforce /etc/apparmor.d/usr.sbin.apache2

    # Restart apache so it loads with the apparmor profiles in enforce mode.
    service apache2 restart
```

A better approach would be to ensure that the apache2 service is stopped before enforcing its profile. Additionally, the method used above (reload, enforce, restart) may or may not accommodate for changes to the AppArmor profiles on-disk, for example when included as part of a planned upgrade. In order to ensure that changes to capabilities or access restrictions are honored, the profile should be explicitly removed and then readded to the running kernel, via `apparmor_parser -R <profile>` and `apparmor_parser -a <profile>`, respectively.  profiles. Again, the service should be stopped before performing these actions.

While investigating this issue, I found a few places where our apache profile needs more attention. I'll open a separate issue to track that discussion.

**Comment by conorsch** [comment]
So far so good on stopping the service rather than reloading it. I've gotten the fix written and have a bit more testing to do. Will open a PR targeting `release/0.3.5` to address.

**Comment by garrettr** [comment]
Closed in #1093.

Github side note/FYI: @conorsch used the auto-close syntax in the description of #1093 but not in the commit message. Since #1093 targeted release/0.3.5, it did not trigger the auto-close when it was merged because Github only autocloses when a PR is merged into the "default branch" (in our case, develop). I cherry-picked the commit from #1093 into develop, but since it did not include auto-close syntax it did not trigger auto-close.

This isn't a problem, but it's good to keep in mind for designing future workflows that use multiple branches and also want to leverage Github's auto-close features.
